GCE and AWS provisioners, dynamic provisioning: admins and user can configure zones where PV shall be created

pospispa opened the pull request:
Note: this PR is based on PR #38505 

Zone configuration capabilities for GCE and AWS dynamic provisioners are extended.
User can request in a persistent volume claim that the dynamically provisioned persistent volume will be created in one of the zone(s) specified by the user.

Fixes Trello cards:
- [GCE provisioner, parse pvc.Selector](https://trello.com/c/CyemTzsK/259-finish-gce-provisioner-parse-pvc-selector-dynamic-provision)
- [AWS provisioner, parse pvc.Selector](https://trello.com/c/2XjouSWw/260-finish-aws-provisioner-parse-pvc-selector-dynamic-provision)

```release-note
GCE and AWS dynamic provisioners extension: users can configure zone(s) in which a persistent volume shall be created.
```

cc: @jsafrane

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37497)
<!-- Reviewable:end -->

Commit:
Added func validatePVCSelector

A user wants to specify a Set of regions and zones where a persistent volume (PV) for his/her persistent volume claim (PVC) shall be created. This is specified in the Selector part of a PVC.

The func validatePVCSelector is added to validate that the user input in the PVC in the Selector part is valid.

Commit:
Added func getPVCMatchLabel

A user may configure a zone in which a persistent volume shall be created using matchLabels in the Selector part of a PVC.

That's why the func getPVCMatchLabel(pvc *api.PersistentVolumeClaim, key string) is added.

Commit:
Added func getPVCMatchExpression

A user may configure a zone in which a persistent volume shall/shall not be created using matchExpressions in the Selector part of a PVC.

That's why the getPVCMatchExpression(pvc *api.PersistentVolumeClaim, key string, operator unversioned.LabelSelectorOperator) is added.

Commit:
Users Can Specify in Which GCE Availability Zone a PV Shall Be Created

A user wants to specify a group of GCE availability zone(s) in which a dynamically provisioned persistent volume shall be created.

That's why the func CreateVolume now uses the func volume.CalculateSetOfZones to get a set of zones that satisfy the above requirement.

Commit:
Users Can Specify in Which AWS Availability Zone a PV Shall Be Created

A user wants to specify a group of AWS availability zone(s) in which a dynamically provisioned persistent volume shall be created.

That's why the func CreateVolume now uses the func volume.CalculateSetOfZones to get a set of zones that satisfy the above requirement.

dims commented:
@k8s-bot ok to test

jsafrane commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -79,6 +81,12 @@ func validateParameters(params map[string]string, fldPath *field.Path) field.Err
 		if len(k) < 1 {
 			allErrs = append(allErrs, field.Invalid(fldPath, k, "field can not be empty."))
 		}
+		if k == "zone" {
```
You can't check all StorageClass instances here, there may be an external or flex provisioner that interprets "zone" parameter in a very different way. All checks must be done in volume plugins (and they're already there, I think).

jsafrane commented on pkg/volume/util.go:
```diff
@@ -345,3 +346,42 @@ func Zones2Set(zonesString string) (sets.String, error) {
+	if len(pvc.Spec.Selector.MatchLabels) > 0 {
+		for label := range pvc.Spec.Selector.MatchLabels {
+			if !allowedKeys[label] {
+				return false, fmt.Errorf("Persistent volume claim (%v), matchLabels part: the %v key is not permitted", pvc.Name, label)
```
- there is no need to introduce `Persistent volume claim (%v),` in the error messages, volume controller will add it and it's quite long already. **This applies to all errors reported in all patches in this PR**

- `matchLabels part: the %v key is not permitted` -> `key %q is not permitted in selector.matchLabels` to produce a sentence and emphasize what's the key here. Applies to all errors below in this patch.

jsafrane commented on pkg/volume/util_test.go:
```diff
@@ -338,3 +339,154 @@ func TestZones2Set(t *testing.T) {
+			pvc: v1.PersistentVolumeClaim{
+				ObjectMeta: v1.ObjectMeta{Name: "pvc", Namespace: "foo"},
+				Spec: v1.PersistentVolumeClaimSpec{
+					Selector: &unversioned.LabelSelector{
```
please add a test with `Selector: nil`

jsafrane commented on pkg/volume/util.go:
```diff
@@ -385,3 +385,16 @@ func ValidatePVCSelector(pvc *v1.PersistentVolumeClaim) (bool, error) {
+	if pvc.Spec.Selector == nil {
+		return "", fmt.Errorf("Persistent volume claim (%v) does not contain any matchLabels", pvc.Name)
+	}
+	if zone, ok := pvc.Spec.Selector.MatchLabels[key]; ok {
```
is it always `zone`? Generic `value` could be more appropriate here.

jsafrane commented on pkg/volume/util.go:
```diff
@@ -398,3 +398,44 @@ func GetPVCMatchLabel(pvc *v1.PersistentVolumeClaim, key string) (string, error)
+//             values:
+//               - us-east-3a
+//               - us-east-4a
+// Returns (sets.String{"us-east-1a": sets.Empty{}, "us-east-2a": sets.Empty{}, "us-east-3a": sets.Empty{}, "us-east-4a": sets.Empty{}}, nil)
```
this example is wrong, docs of `MatchExpressions` says that "The requirements are ANDed", therefore it must match both expressions, i.e. only "us-east-3a" should be returned.

And beware of `NotIn` operator, it behaves differently than `In` here.

jsafrane commented on pkg/volume/util.go:
```diff
@@ -439,3 +439,30 @@ func GetPVCMatchExpression(pvc *v1.PersistentVolumeClaim, key string, operator u
 }
+
+// Regions2ZonesParams are parameters for the func Regions2Zones
+type Regions2ZonesParams struct {
```
I think you don't need an extra structure for 3 parameters. Or do I miss something?

jsafrane commented on pkg/volume/util.go:
```diff
@@ -439,3 +439,30 @@ func GetPVCMatchExpression(pvc *v1.PersistentVolumeClaim, key string, operator u
+	// all currently available zones
+	AllAvailableZones sets.String
+	// a func that converts a zone to a region
+	Zone2region func(string) (string, error)
```
why do you return an error here if you throw it away in `Regions2Zones`?

jsafrane reviewed (commented):
halfway through the patches, will continue tomorrow

jsafrane reviewed (commented):
From failed verification test:

```
I1126 15:01:54.005] BUILD files are not up to date
I1126 15:01:54.006] Run ./hack/update-bazel.sh
I1126 15:01:54.014] FAILED   hack/make-rules/../../hack/verify-bazel.sh	23s
```

pospispa commented on pkg/volume/util.go:
```diff
@@ -331,3 +332,215 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+// GetPVCMatchLabel returns:
+// - either (value, nil) for the key from the matchLabels Selector part of the PVC
+// - or ("", error) in case the key is missing in the matchLabels Selector part of the PVC
+func GetPVCMatchLabel(pvc *v1.PersistentVolumeClaim, key string) (string, error) {
```
I agree that simple bool will do, however, IMHO, to return an error is a more general solution that can be used easily in future.

justinsb replied:
And there are already two different errors, so it feels like an error is right, I agree

jsafrane commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -91,12 +92,19 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (aws.K
 		case "zone":
-			volumeOptions.AvailabilityZone = v
+			zonePresent = true
+			calculateZonesParams.StorageClassZones = v
```
This will interpret `zone` parameter as comma separated list of zones, right? IMO it's not what we want.

pospispa replied:
Yes, it will interpret the `zone` parameter as a comma separated list of zones. Assume, the documentation won't mention this extra functionality. Is it a problem considering the `zone` parameter will be removed in (near) future because it is being replaced by the `zones` parameter?

jsafrane replied:
It won't be "near future", this is API change and it must be working until Kubernetes 2.0, whenever it is released.

justinsb commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -114,11 +122,20 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (aws.K
+	calculateZonesParams.IsSCZoneSpecified = zonePresent || zonesPresent
+	calculateZonesParams.PVC = c.options.PVC
+	calculateZonesParams.GetAllZones = cloud.GetAllZones
+	calculateZonesParams.Zone2region = aws.AzToRegion
```
I think our style is to spell it out i.e. `ZoneToRegion`

justinsb commented on pkg/volume/util.go:
```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
 }
+
+// Zones2Set converts a string containing a comma separated list of zones to set
+func Zones2Set(zonesString string) (sets.String, error) {
```
ZonesToSet

justinsb commented on pkg/volume/util.go:
```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+	for _, zone := range zonesSlice {
+		trimmedZone := strings.TrimSpace(zone)
+		if trimmedZone == "" {
+			return make(sets.String), fmt.Errorf("comma separated list of zones must not contain an empty zone")
```
You could probably include the list of zones, i.e.

`"comma separated list of zones (%q) must not contain an empty zone", zonesString`

justinsb commented on pkg/volume/util.go:
```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+
+// Regions2Zones converts a set of regions to a set of zones
+func Regions2Zones(in Regions2ZonesParams) (sets.String, error) {
+	ret := make(sets.String)
```
Might be clearer to name this `regions`, but that's just nit-picking

justinsb commented on pkg/volume/util.go:
```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+}
+
+// Regions2Zones converts a set of regions to a set of zones
+func Regions2Zones(in Regions2ZonesParams) (sets.String, error) {
```
Is there a reason to use the class-as-argument pattern here?  It feels surprising...

justinsb commented on pkg/volume/util.go:
```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+	for zone := range in.AllAvailableZones {
+		for region := range in.Regions {
+			if regionOfZone, err := in.Zone2region(zone); err != nil {
+				return make(sets.String), fmt.Errorf("failed to convert zone (%v) to a region: %v", zone, err.Error())
```
I think just `err`, not `err.Error()`

justinsb commented on pkg/volume/util.go:
```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+		for region := range in.Regions {
+			if regionOfZone, err := in.Zone2region(zone); err != nil {
+				return make(sets.String), fmt.Errorf("failed to convert zone (%v) to a region: %v", zone, err.Error())
+			} else if strings.Compare(regionOfZone, region) == 0 {
```
Is this the same as `regionOfZone == region`?

I'm actually very unclear as to what this whole function is actually doing.

Can we just call ZoneToRegion once per zone, and then check if in.Regions `Has` the region?

I could use a comment as to why we're doing this per-region filtering!

justinsb commented on pkg/volume/util.go:
```diff
@@ -332,3 +332,239 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+	var ret, allAvailableZones sets.String
+	var err error
+	emptySet := make(sets.String)
+	if allAvailableZones, err = input.GetAllZones(); err != nil {
```
I think if the user has specified zones in their storage class, we should not look at GetAllZones.

justinsb reviewed (changes requested):
Generally like the approach, but some tweaks needed IMO

pospispa commented:
@k8s-bot verify test this

pospispa commented:
@justinsb I believe, I have addressed all your comment so PTAL.

justinsb commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -91,12 +92,17 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (aws.K
 	}
 	// Apply Parameters (case-insensitive). We leave validation of
 	// the values to the cloud provider.
+	var PutAdminAndUserRequestsTogetherParams volume.PutAdminAndUserRequestsTogetherParams
```
Style nit - I think the variable name should start with a lower-case letter i.e. `var putAdminAnd...`

justinsb commented on pkg/volume/gce_pd/gce_util.go:
```diff
@@ -86,33 +86,30 @@ func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (strin
 	// the values to the cloud provider.
 	diskType := ""
-	zone := ""
+	var PutAdminAndUserRequestsTogetherParams volume.PutAdminAndUserRequestsTogetherParams
```
Lower case here also `var putAdmin...`

justinsb commented on pkg/volume/gce_pd/gce_util.go:
```diff
@@ -86,33 +86,30 @@ func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (strin
 			diskType = v
 		case "zone":
-			zone = v
+			PutAdminAndUserRequestsTogetherParams.IsSCZoneSpecified = true
```
Does this get cleaner if we pass `c.options` in to `volume.PutAdminAndUserRequestsTogether` ?

pospispa replied:
The `c.options` are validated and lower-cased in the `func CreateVolume`. If we put `c.options` into `volume.PutAdminAndUserRequestsTogether` the `for k, v := range c.options.Parameters` will have to be repeated inside the `func` and the  `c.options.Parameters` will have to be lower-cased again.
Note: I wouldn't put the validation of the `c.options.Parameters` into the `func volume.PutAdminAndUserRequestsTogether` because the `func` deals with `zone, zones and PVC` only.
On the other hand if the `func volume.PutAdminAndUserRequestsTogether` is made a class it might get a little bit better.

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+// - (false, nil) means PVC is valid (error == nil) and there is at least a value in matchLabels or matchExpressions specified (bool == false)
+// - (false, error) means PVC is not valid
+// - (true, error) shall never happen
+func ValidatePVCSelector(pvc *v1.PersistentVolumeClaim) (bool, error) {
```
I don't think we need to export this function i.e. `validatePVCSelector`

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
 }
+
+// ZonesToSet converts a string containing a comma separated list of zones to set
+func ZonesToSet(zonesString string) (sets.String, error) {
```
I do not think this needs to be exported (`zonesToSet`)

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+	return false, nil
+}
+
+// GetPVCMatchLabel returns:
```
I do not think this needs to be exported - `getPVCMatchLabel`

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+	return "", fmt.Errorf("key %q not found in selector.matchLabels", key)
+}
+
+// GetPVCMatchExpression returns:
```
Does not need to be exported (?)

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+	zones := make(sets.String)
+	for candidate := range allAvailableZones {
+		if candidatesRegion, err := zoneToRegion(candidate); err != nil {
+			return make(sets.String), fmt.Errorf("failed to convert zone (%v) to a region: %v", candidate, err)
```
I believe you can return `nil, fmt.Errorf...` ?

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+	return ret, nil
+}
+
+// RegionsToZones converts a set of regions to a set of zones
```
So AFAICT this is only called with the set of zones as returned by GetAllZones.

How about having GetAllZones either return `map[string]string` mapping zones to regions, or (preferred IMO), `ZoneInfo[]` where ZoneInfo includes the Zone & Region?

pospispa replied:
Well, in case an admin doesn't configure any zone(s) in the StorageClass "the default" is a set of all zones that's why the current implementation of the `func GetAllZones` is useful from my point of view.
So I would personally leave the `func GetAllZones` as is.
However, it may be useful to create map[region]SetOfZones data structure and use it for Regions -> SetOfZones conversion instead using the current implementation of the `func RegionsToZones`.

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+	gotAllAvailableZones := false
+	emptySet := make(sets.String)
+	if input.IsSCZoneSpecified && input.IsSCZonesSpecified {
+		return emptySet, fmt.Errorf("both zone and zones StorageClass parameters must not be used at the same time")
```
I think just return nil instead of emptySet, when returning an error

justinsb commented on pkg/volume/util.go:
```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+// PutAdminAndUserRequestsTogether returns:
+// - either a set of zones resulting from currently available zones, allowed zone(s) by an admin in the corresponding storage class and zones preferred by the user in the selector part of the PVC
+// - or an error in case the resulting set of zones is empty or another error occurred
+func PutAdminAndUserRequestsTogether(input PutAdminAndUserRequestsTogetherParams) (sets.String, error) {
```
It might actually be easier to make this function a class, given the complex arguments & state (e.g. gotAllAvailableZones).

justinsb reviewed (commented):
I think the code logic is solid here, just some style points

Commit:
func ZonesToSet doesn't have to be public any longer.

Commit:
Added class ZonesConf

There is or may be:
  - Set of currently available zones.
  - Comma separated set of zone(s) configured by an admin in a StorageClass.
  - User requested or unwanted sets of zones and regions in a PVC in the Selector part.
It's necessary to put all the above information together and calculate a set of zones in which the requested PV can be created.

That's why the class ZonesConf is added.

pospispa commented:
@justinsb I believe, I have addressed all your comment so PTAL.

jsafrane commented:
all review remarks were addressed
/lgtm
/approve

@thockin, @zmerlynn please approve changes in examples/
@justinsb, please approve AWS changes

justinsb commented:
@k8s-bot unit test this

/approve

/lgtm

pospispa commented:
@k8s-bot cvm gke e2e test this

pospispa commented:
@k8s-bot bazel test this

pospispa commented:
@k8s-bot cvm gce e2e test this

Commit:
func ValidateZone doesn't have to be public any longer.

pospispa commented:
@k8s-bot node e2e test this

pospispa commented:
@k8s-bot kops aws e2e test this

marun commented:
Moving out of v1.8, move back if required.

pospispa commented:
/test pull-kubernetes-bazel

pospispa commented:
/test pull-kubernetes-unit

pospispa commented:
@k8s-bot kops aws e2e test this

pospispa commented:
It was found out during review discussions of the related PR https://github.com/kubernetes/kubernetes/pull/38505 that this PR requires a design.
The design was created and discussed in https://github.com/kubernetes/community/pull/247
However, the design discussions ended up in lack of use-cases and that's why the design was never approved and was closed recently.
That's why I'm closing this PR.
